File: android/app/src/main/java/com/noop/oura/RecordInventory.kt
package com.noop.oura

import java.util.Locale

/**
 * Privacy-safe metadata for complete Oura history records observed by the decoder.
 *
 * This deliberately retains only the tag, record count, wire-byte count, and current typed-event
 * count. It never retains record payloads, timestamps, device identifiers, authentication material,
 * or biometric values, so [summary] is safe for NOOP's normal redacted diagnostic log.
 */
class OuraRecordInventory {
    data class Entry(
        val tag: Int,
        val recordCount: Int,
        val wireByteCount: Int,
        val emittedEventCount: Int,
        val decodedRecordCount: Int,
    ) {
        /** The current NOOP dictionary entry, or null for a genuinely unknown tag. */
        val knownTag: OuraEventTag? get() = OuraEventTag.fromRaw(tag)

        /** Structurally valid records that intentionally or unexpectedly emitted no typed event. */
        val silentRecordCount: Int get() = recordCount - decodedRecordCount
    }

    private data class MutableEntry(
        var recordCount: Int = 0,
        var wireByteCount: Int = 0,
        var emittedEventCount: Int = 0,
        var decodedRecordCount: Int = 0,
    )

    private val counts = mutableMapOf<Int, MutableEntry>()

    /** Observe one complete, structurally-valid TLV after the current decoder has processed it. */
    fun observe(record: OuraRecord, emittedEventCount: Int) {
        val entry = counts.getOrPut(record.type) { MutableEntry() }
        entry.recordCount += 1
        entry.wireByteCount += record.totalLength
        entry.emittedEventCount += emittedEventCount.coerceAtLeast(0)
        if (emittedEventCount > 0) entry.decodedRecordCount += 1
    }

    /** Stable tag-ascending entries for deterministic logs and tests. */
    val entries: List<Entry>
        get() = counts.keys.sorted().map { tag ->
            val value = requireNotNull(counts[tag])
            Entry(
                tag = tag,
                recordCount = value.recordCount,
                wireByteCount = value.wireByteCount,
                emittedEventCount = value.emittedEventCount,
                decodedRecordCount = value.decodedRecordCount,
            )
        }

    val totalRecordCount: Int get() = counts.values.sumOf { it.recordCount }
    val totalWireByteCount: Int get() = counts.values.sumOf { it.wireByteCount }
    val emittedEventCount: Int get() = counts.values.sumOf { it.emittedEventCount }
    val unknownRecordCount: Int
        get() = entries.filter { it.knownTag == null }.sumOf { it.recordCount }
    val isEmpty: Boolean get() = counts.isEmpty()

    /**
     * One bounded, deterministic, values-free line. Each item is `tag(name)=records/events/bytes`; a tag absent
     * from the current dictionary is named `UNKNOWN`.
     */
    fun summary(maximumTags: Int = 24): String {
        val allEntries = entries
        val visible = allEntries.take(maximumTags.coerceAtLeast(0))
        val parts = visible.map { entry ->
            val tag = String.format(Locale.ROOT, "0x%02X", entry.tag)
            val name = entry.knownTag?.tagName ?: "UNKNOWN"
            "$tag($name)=${entry.recordCount}r/${entry.emittedEventCount}e/${entry.wireByteCount}b"
        }
        val omitted = (allEntries.size - visible.size).coerceAtLeast(0)
        val suffix = if (omitted > 0) " +$omitted tag(s)" else ""
        val detail = if (parts.isEmpty()) "none$suffix" else parts.joinToString(",") + suffix
        return "records=$totalRecordCount bytes=$totalWireByteCount events=$emittedEventCount " +
            "unknown=$unknownRecordCount tags=[$detail]"
    }
}

File: Packages/OuraProtocol/Sources/OuraProtocol/RecordInventory.swift
import Foundation

/// A privacy-safe inventory of complete Oura history records observed by the decoder.
///
/// The inventory deliberately retains metadata only: tag, record count, wire-byte count, and the
/// number of typed events emitted by NOOP's current decoder. It never retains record payloads,
/// timestamps, device identifiers, auth material, or biometric values. This makes the summary safe to
/// include in the normal redacted strap log while still answering the protocol-mapping questions that
/// matter: which tags a firmware emitted, whether NOOP knows their names, and whether they currently
/// produce typed events.
public struct OuraRecordInventory: Equatable, Sendable {
    public struct Entry: Equatable, Sendable {
        public let tag: UInt8
        public let recordCount: Int
        public let wireByteCount: Int
        public let emittedEventCount: Int
        public let decodedRecordCount: Int

        public init(tag: UInt8, recordCount: Int, wireByteCount: Int,
                    emittedEventCount: Int, decodedRecordCount: Int) {
            self.tag = tag
            self.recordCount = recordCount
            self.wireByteCount = wireByteCount
            self.emittedEventCount = emittedEventCount
            self.decodedRecordCount = decodedRecordCount
        }

        /// The current NOOP dictionary entry, or nil when this is a genuinely unknown tag.
        public var knownTag: OuraEventTag? { OuraEventTag(rawValue: tag) }

        /// Records from this tag that were structurally valid but produced no typed event. A known
        /// lifecycle/diagnostic tag may intentionally be silent, so this is not labelled a decode error.
        public var silentRecordCount: Int { recordCount - decodedRecordCount }
    }

    private struct MutableEntry: Equatable, Sendable {
        var recordCount = 0
        var wireByteCount = 0
        var emittedEventCount = 0
        var decodedRecordCount = 0
    }

    private var counts: [UInt8: MutableEntry] = [:]

    public init() {}

    /// Observe one complete, structurally-valid TLV record after the current decoder has processed it.
    /// Passing the event count keeps the inventory independent of app/storage types and lets one record
    /// honestly report that it expanded to several IBI/temperature/etc. samples.
    public mutating func observe(_ record: OuraRecord, emittedEventCount: Int) {
        var entry = counts[record.type] ?? MutableEntry()
        entry.recordCount += 1
        entry.wireByteCount += record.totalLength
        entry.emittedEventCount += max(0, emittedEventCount)
        if emittedEventCount > 0 { entry.decodedRecordCount += 1 }
        counts[record.type] = entry
    }

    /// Stable tag-ascending entries for deterministic logs, CLI output, and tests.
    public var entries: [Entry] {
        counts.keys.sorted().compactMap { tag in
            guard let value = counts[tag] else { return nil }
            return Entry(tag: tag,
                         recordCount: value.recordCount,
                         wireByteCount: value.wireByteCount,
                         emittedEventCount: value.emittedEventCount,
                         decodedRecordCount: value.decodedRecordCount)
        }
    }

    public var totalRecordCount: Int { counts.values.reduce(0) { $0 + $1.recordCount } }
    public var totalWireByteCount: Int { counts.values.reduce(0) { $0 + $1.wireByteCount } }
    public var emittedEventCount: Int { counts.values.reduce(0) { $0 + $1.emittedEventCount } }

    /// Count of records whose tag is absent from NOOP's dictionary. This is the honest mapping backlog;
    /// known-but-silent records are kept separate because some lifecycle tags intentionally emit nothing.
    public var unknownRecordCount: Int {
        entries.filter { $0.knownTag == nil }.reduce(0) { $0 + $1.recordCount }
    }

    public var isEmpty: Bool { counts.isEmpty }

    /// One bounded, deterministic, values-free line suitable for NOOP's exportable strap log.
    /// Each item is `tag(name)=records/events/bytes`; an unknown tag is named `UNKNOWN`.
    public func summary(maximumTags: Int = 24) -> String {
        let limit = max(0, maximumTags)
        let visible = entries.prefix(limit)
        let parts = visible.map { entry in
            let tag = String(format: "0x%02X", entry.tag)
            let name = entry.knownTag?.name ?? "UNKNOWN"
            return "\(tag)(\(name))=\(entry.recordCount)r/\(entry.emittedEventCount)e/\(entry.wireByteCount)b"
        }
        let omitted = max(0, entries.count - visible.count)
        let suffix = omitted > 0 ? " +\(omitted) tag(s)" : ""
        let detail = parts.isEmpty ? "none" + suffix : parts.joined(separator: ",") + suffix
        return "records=\(totalRecordCount) bytes=\(totalWireByteCount) "
            + "events=\(emittedEventCount) unknown=\(unknownRecordCount) tags=[\(detail)]"
    }
}
